### PR TITLE
feat(TCOMP-2085): record builder entries ordering

### PR DESCRIPTION
**Pull request by undx**
https://jira.talendforge.org/browse/TCOMP-2085

### CODE REVIEW

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java [hunk 1]
```diff
@@ -454,6 +463,69 @@ private void assertType(final Schema.Type actual, final Schema.Type expected) {
...
+        private Order orderState = Order.LAST;
+
+        private String orderTarget = "";
+
+        // flag if providedSchema's entriesOrder was altered
+        private Boolean orderOverride = Boolean.FALSE;
+
+        private List<String> entriesOrder = new ArrayList<>();
+
+        private void updateOrderState(final String name) {
```
Comment (clesaec) [inline component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java:497]: Could you put this in external class, something like :
```java
class NamePosition {
    enum Order { BEFORE, AFTER, LAST;  }
    private final Order orderState;
     private final String orderTarget;

     public void updateOrderState(final String newName, List<String> names) { ... }
}
```
(Not mandatory,  but easier to test and to read.)

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/SchemaImpl.java [hunk 2]
```diff
@@ -308,7 +310,20 @@ public Builder swap(final String name, final String with) {
...
         @Override
         public Schema build() {
-            this.props.put(ENTRIES_ORDER_PROP, entriesOrder.stream().collect(Collectors.joining(",")));
+            this.props.put(ENTRIES_ORDER_PROP, entriesOrder.stream().collect(joining(",")));
+            return new SchemaImpl(this);
+        }
+
+        @Override
+        public Schema build(final EntriesOrder order) {
+            final List<String> existing = entries.stream().map(e -> e.getName()).collect(toList());
```
Comment (clesaec) [inline component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/SchemaImpl.java:319]: This code should be enough
```java
            final String propEntriesOrder = entries.stream().sorted(order).map(Entry::getName).collect(joining(",");
            this.props.put(ENTRIES_ORDER_PROP, propEntriesOrder);
```
and `public Schema build(final EntriesOrder order) {` can be replaced with `public Schema build(final Comparator<Entry> order) {`

**Review by clesaec — CHANGES_REQUESTED** [review]
Some comments and changes required.

File: component-api/src/main/java/org/talend/sdk/component/api/record/Schema.java [hunk 3]
```diff
@@ -431,6 +431,17 @@ default Builder swap(final String name, final String with) {
...
         Schema build();
+
+        /**
+         * Same as {@link Builder#build()} but entries order is specified by {@code order}. This takes precedence on any
+         * previous defined order with builder and may void it.
+         *
+         * @param order the wanted order for entries.
+         * @return the described schema.
+         */
+        default Schema build(EntriesOrder order) {
```
[the remark this answers is not in the document]
Comment (undx) [reply component-api/src/main/java/org/talend/sdk/component/api/record/Schema.java:442]: Yep, sure!

File: component-runtime-beam/src/main/java/org/talend/sdk/component/runtime/beam/spi/record/AvroSchemaBuilder.java [hunk 4]
```diff
@@ -339,7 +347,20 @@ public Schema build() {
...
-            record.addProp(ENTRIES_ORDER_PROP, fields.stream().map(e -> e.getName()).collect(Collectors.joining(",")));
+            if (order != null) {
+                final List<String> existing = fields.stream().map(e -> e.getName()).collect(toList());
+                // sanity check: filter non-existing entries passed to builder
+                List<String> merge =
+                        order.getFieldsOrder().stream().filter(o -> existing.contains(o)).collect(toList());
+                // keep entries created w/ builder and not passed w/ order
+                existing.removeAll(order.getFieldsOrder());
+                // order has precedence so we add missing order entries at the end
+                merge.addAll(existing);
```
[the remark this answers is not in the document]
Comment (undx) [reply component-runtime-beam/src/main/java/org/talend/sdk/component/runtime/beam/spi/record/AvroSchemaBuilder.java:358]: more simpler for same effect ;-)

**Review by clesaec — DISMISSED** [review]
LGTM

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java (hunk 1 above)
Comment (undx) [reply component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java]: done

**Review by clesaec — COMMENTED** [review]
:+1:

**Review by clesaec — APPROVED** [review]

**Pull request merged**